File: General-DSA-Questions/Stack/5_nextGreaterElement.cpp
// Given an array arr[] of size n, for each element arr[i], find the next greater element to the right.
// If no such element exists, return -1.

// Notes : https://takeuforward.org/data-structure/next-greater-element-using-stack/

// Intuition: (Use Monotonic Decreasing Stack : top of stack has the smallest of the larger elements seen so far)
// 1. You process from right to left so that you’ve already seen future values.
// 2. You use the stack to hold possible candidates for the next greater element.
// 3. You pop smaller elements because they’re useless for future elements.
//    a. The main intuition behind popping them is that these elements can never be the NGE for any element present at the left of A[i] because A[i] is greater than these elements.
// 4. The top of the stack is always the next greater element for the current element.

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

vector<int> nextGreaterElements(const vector<int>& arr) {
    int n = arr.size();
    vector<int> result(n, -1);  // Initialize result with -1
    stack<int> s;

    for (int i = n - 1; i >= 0; --i) {
        // Pop all elements smaller than or equal to current
        while (!s.empty() && s.top() <= arr[i]) {
            s.pop();
        }

        if (!s.empty()) { // If stack is not empty, the top element is the next greater element
            result[i] = s.top();
        }

        s.push(arr[i]); // Push the current element onto the stack
    }

    return result;
}

int main() {
    vector<int> arr = {4, 5, 2, 10, 8};
    vector<int> res = nextGreaterElements(arr);

    cout << "Next Greater Elements:\n";
    for (int val : res) {
        cout << val << " ";
    }

    return 0;
}


// Time Complexity: O(n) - Each element is pushed and popped from the stack at most once.
// Space Complexity: O(n) - The stack can hold up to n elements in the worst case.

File: JS-DSA-Questions/Stack/5_nextGreaterElement.js
// Given an array arr[] of size n, for each element arr[i], find the next greater element to its
// right. If none exists, use -1.

// Notes : https://takeuforward.org/data-structure/next-greater-element-using-stack/

// Intuition: Use a monotonic decreasing stack (top holds the smallest of the larger elements seen).
// 1. Process from right to left so future values are already seen.
// 2. The stack holds candidates for the next greater element.
// 3. Pop elements <= current, since they can never be the NGE for anything to the left.
// 4. The stack top (if any) is the next greater element for the current element.

// JavaScript note: a plain array is the stack (push()/pop(), arr[arr.length-1] as top).
// Array(n).fill(-1) is the result initialized to -1.
function nextGreaterElements(arr) {
  const n = arr.length;
  const result = new Array(n).fill(-1);
  const s = [];

  for (let i = n - 1; i >= 0; i--) {
    // Pop all elements smaller than or equal to the current
    while (s.length > 0 && s[s.length - 1] <= arr[i]) {
      s.pop();
    }

    if (s.length > 0) {
      result[i] = s[s.length - 1];
    }

    s.push(arr[i]);
  }

  return result;
}

console.log("Next Greater Elements:", nextGreaterElements([4, 5, 2, 10, 8]).join(" ")); // 5 10 -1 -1 -1

// Time Complexity: O(n). Space Complexity: O(n).
